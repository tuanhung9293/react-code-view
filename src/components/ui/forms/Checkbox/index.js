import React from 'react'
import styles from './Checkbox.module.css'

function Checkbox({hasError, ...props}) {
  return (
    <span className={styles.wrapper}>
      <input {...props} className={styles.input} type="checkbox" />
      <span className={styles.display} />
    </span>
  )
}

export default Checkbox