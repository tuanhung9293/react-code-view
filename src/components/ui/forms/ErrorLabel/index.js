import React from 'react'
import styles from './ErrorLabel.module.css'

function ErrorLabel({...rest}) {
  return <label className={styles.base} {...rest} />
}

export default ErrorLabel