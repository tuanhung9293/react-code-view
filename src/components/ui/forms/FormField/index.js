import React from 'react'
import {connect} from 'react-redux'
import PropTypes from 'prop-types'
import FormField from './FormField'
import {setFormValue, formFieldBlur, validateField} from 'state/actions/formsAction'

export function mapStateToProps({forms}, {formName, name}) {
  const form = forms[formName] || {values: {}, initialValues: {}, touched: {}, validationErrors: {}}
  const value = form.values.hasOwnProperty(name) ? form.values[name] : form.initialValues[name]
  return {
    value: value,
    validationError: form.validationErrors[name] || null,
    hasError: !!(form.touched[name] && form.validationErrors[name]),
    touched: form.touched[name] || false
  }
}

export function mapDispatchToProps(dispatch, {formName, name}) {
  return {
    onChange: ({target: {value}}) => dispatch(setFormValue({form: formName, name, value})),
    onBlur: () => {
      dispatch(formFieldBlur({form: formName, name}))
    },
    validateField: (error) => dispatch(validateField({form: formName, name, error}))
  }
}

const ConnectedFormField = connect(mapStateToProps, mapDispatchToProps)(FormField)

function ConnectedFormFieldWithContext(props, context) {
  const formName = context._form && context._form.name
  return <ConnectedFormField {...props} formName={formName} />
}

ConnectedFormFieldWithContext.contextTypes = {
  _form: PropTypes.object
}

export default ConnectedFormFieldWithContext