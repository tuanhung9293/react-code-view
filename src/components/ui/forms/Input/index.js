import React from 'react'
import classNames from 'classnames'
import styles from './Input.module.css'

function Input({type, hasError, value, ...rest}) {
  const inputValue = value || ''
  const className = classNames({
    [styles.input]: true,
    [styles.error]: hasError
  })
  return (
    <input className={className} type={type || 'text'} value={inputValue} {...rest} />
  )
}

export default Input