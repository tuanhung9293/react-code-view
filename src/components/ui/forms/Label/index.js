import React from 'react'
import classNames from 'classnames'
import styles from './Label.module.css'

function Label({children, title}) {
  const className = classNames({
    [styles.base]: true,
    [styles.title]: title
  })
  return (
    <label className={className}>{children}</label>
  )
}

export default Label