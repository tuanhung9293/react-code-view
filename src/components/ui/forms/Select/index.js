import React from 'react'
import classNames from 'classnames'
import styles from './Select.module.css'

function Select({hasError, value, slim, ...props}) {
  const className = classNames({
    [styles.select]: true,
    [styles.error]: hasError,
    [styles.slim]: slim
  })
  return (
    <div className={styles.wrapper}>
      <select value={value || ''} {...props} className={className} />
    </div>
  )
}

export default Select