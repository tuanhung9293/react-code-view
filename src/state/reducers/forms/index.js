import {
  SET_INITIAL_FORM_VALUES,
  FORM_FIELD_BLUR,
  VALIDATE_FORM,
  VALIDATE_FIELD,
  SET_FORM_VALUE,
  CLEAR_FORM
} from 'constants/actionTypes'

const initialFormState = {
  meta: {
    submitting: false,
    submitted: false
  },
  values: {},
  initialValues: {},
  validationErrors: {},
  touched: {}
}

export function form(form = initialFormState, action) {
  switch (action.type) {
    case SET_INITIAL_FORM_VALUES:
      return {
        ...form,
        initialValues: {
          ...form.initialValues,
          ...action.payload.initialValues
        }
      }
    case FORM_FIELD_BLUR:
      return {
        ...form,
        touched: {
          ...form.touched,
          [action.payload.name]: true
        }
      }
    case VALIDATE_FORM:
      return {
        ...form,
        validationErrors: {...action.payload.errors}
      }
    case VALIDATE_FIELD:
      if (action.payload.error) {
        return {
          ...form,
          validationErrors: {
            ...form.validationErrors,
            [action.payload.name]: action.payload.error
          }
        }
      } else {
        let nextValidationErrors = {...form.validationErrors}
        delete nextValidationErrors[action.payload.name]
        return {
          ...form,
          validationErrors: nextValidationErrors
        }
      }
    case SET_FORM_VALUE:
      return {
        ...form,
        values: {
          ...form.values,
          [action.payload.name]: action.payload.value
        }
      }
    default:
      return form
  }
}

function forms(forms = {}, action) {
  switch (action.type) {
    case SET_INITIAL_FORM_VALUES:
    case SET_FORM_VALUE:
    case VALIDATE_FORM:
    case VALIDATE_FIELD:
    case FORM_FIELD_BLUR:
      return {
        ...forms,
        [action.payload.form]: form(forms[action.payload.form], action)
      }
    case CLEAR_FORM:
      let clone = {...forms}
      delete clone[action.payload.form]
      return clone
    default:
      return forms
  }
}

export default forms