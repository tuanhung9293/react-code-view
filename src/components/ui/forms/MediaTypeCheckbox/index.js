
import React from 'react'
import Checkbox from 'components/ui/forms/Checkbox'
import Label from 'components/ui/forms/Label'
import styles from './MediaTypeCheckbox.module.css'

function MediaTypeCheckbox({video, image, audio, favorite, onImageCheck, onVideoCheck, onAudioCheck, onFavoriteCheck, label}) {
  return (
    <div className={styles.wrapper}>
      {label &&
        <div className={styles.labels}>
          <Label>{label}</Label>
        </div>
      }
      <div className={styles.inlineFlex}>
        <label className={styles.checkbox}>
          <Checkbox checked={image} onChange={onImageCheck} />Image
        </label>
        <label className={styles.checkbox}>
          <Checkbox checked={video} onChange={onVideoCheck} />Video
        </label>
        <label className={styles.checkbox}>
          <Checkbox checked={audio} onChange={onAudioCheck} />Audio
        </label>
        {favorite !== undefined &&
          <label className={styles.checkbox}>
            <Checkbox checked={favorite} onChange={onFavoriteCheck} />Favorite
          </label>
        }
      </div>
    </div>
  )
}

export default MediaTypeCheckbox