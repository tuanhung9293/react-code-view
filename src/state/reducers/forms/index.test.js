import forms, {form} from './'
import {
  VALIDATE_FORM,
  CLEAR_FORM,
  VALIDATE_FIELD,
  FORM_SUBMIT
} from 'constants/actionTypes'

describe('forms', () => {
  it('has initial state', () => {
    expect(forms(undefined, {})).toMatchSnapshot()
  })

  it('responds to VALIDATE_FORM', () => {
    expect(forms(undefined, {
      type: VALIDATE_FORM,
      payload: {
        form: 'testForm',
        errors: {testField: 'is required'}
      }
    })).toMatchSnapshot()
  })

  it('responds to VALIDATE_FIELDS', () => {
    expect(forms(undefined, {
      type: VALIDATE_FIELD,
      payload: {
        form: 'testForm',
        name: 'testField',
        error: 'is required'
      }
    })).toMatchSnapshot()
  })

  it('clears errors on VALIDATE_FIELD', () => {
    const state = forms(undefined, {
      type: VALIDATE_FIELD,
      payload: {
        form: 'testForm',
        name: 'testField',
        error: 'is required'
      }
    })
    expect(forms(state, {
      type: VALIDATE_FIELD,
      payload: {
        form: 'testForm',
        name: 'testField',
        error: false
      }
    })).toMatchSnapshot()
  })

  it('deletes the form on CLEAR_FORM', () => {
    const state = {
      testForm: {
        values: {},
        errors: {}
      }
    }
    expect(forms(state, {
      type: CLEAR_FORM,
      payload: {
        form: 'testForm'
      }
    })).toMatchSnapshot()
  })
})

describe('form', () => {
  it('has initial state', () => {
    expect(form(undefined, {})).toMatchSnapshot()
  })
})
