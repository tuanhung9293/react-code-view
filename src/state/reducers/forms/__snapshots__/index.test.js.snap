// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form has initial state 1`] = `
Object {
  "initialValues": Object {},
  "meta": Object {
    "submitted": false,
    "submitting": false,
  },
  "touched": Object {},
  "validationErrors": Object {},
  "values": Object {},
}
`;

exports[`forms clears errors on VALIDATE_FIELD 1`] = `
Object {
  "testForm": Object {
    "initialValues": Object {},
    "meta": Object {
      "submitted": false,
      "submitting": false,
    },
    "touched": Object {},
    "validationErrors": Object {},
    "values": Object {},
  },
}
`;

exports[`forms deletes the form on CLEAR_FORM 1`] = `Object {}`;

exports[`forms has initial state 1`] = `Object {}`;

exports[`forms responds to VALIDATE_FIELDS 1`] = `
Object {
  "testForm": Object {
    "initialValues": Object {},
    "meta": Object {
      "submitted": false,
      "submitting": false,
    },
    "touched": Object {},
    "validationErrors": Object {
      "testField": "is required",
    },
    "values": Object {},
  },
}
`;

exports[`forms responds to VALIDATE_FORM 1`] = `
Object {
  "testForm": Object {
    "initialValues": Object {},
    "meta": Object {
      "submitted": false,
      "submitting": false,
    },
    "touched": Object {},
    "validationErrors": Object {
      "testField": "is required",
    },
    "values": Object {},
  },
}
`;
